src/SignalR/clients/cpp/src/signalrclient/websocket_transport.cpp

```diff
@@ -41,7 +43,99 @@ namespace signalr
+        // incremented when the shared pointer is acquired and then decremented when it goes out of scope of the continuation.
+        auto weak_transport = std::weak_ptr<websocket_transport>(this_transport);
+
+        auto websocket_client = this_transport->safe_get_websocket_client();
+
+        // There are two cases when we exit the loop. The first case is implicit - we pass the cancellation_token
+        // to `then` (note this is after the lambda body) and if the token is cancelled the continuation will not
+        // run at all. The second - explicit - case happens if the token gets cancelled after the continuation has
+        // been started in which case we just stop the loop by not scheduling another receive task.
+        websocket_client->receive([weak_transport, cts, logger, websocket_client](std::string message, std::exception_ptr exception)
+            {
+                if (exception != nullptr)
+                {
+                    try
+                    {
+                        std::rethrow_exception(exception);
+                    }
+                    catch (const std::exception & e)
+                    {
+                        cts.cancel();
+
+                        logger.log(
+                            trace_level::errors,
+                            std::string("[websocket transport] error receiving response from websocket: ")
+                            .append(e.what()));
+
+                        websocket_client->stop([](std::exception_ptr exception)
+                            {
+                            });
+
+                        auto transport = weak_transport.lock();
+                        if (transport)
+                        {
+                            transport->m_close_callback(exception);
+                        }
+                    }
+                    catch (...)
+                    {
+                        cts.cancel();
+
+                        logger.log(
+                            trace_level::errors,
+                            std::string("[websocket transport] unknown error occurred when receiving response from websocket"));
+
+                        websocket_client->stop([](std::exception_ptr)
+                            {
+                            });
+
+                        auto transport = weak_transport.lock();
+                        if (transport)
+                        {
+                            transport->m_close_callback(std::make_exception_ptr(signalr_exception("unknown error")));
+                        }
+                    }
+
+                    return;
+                }
+
+                auto transport = weak_transport.lock();
+                if (transport)
+                {
+                    transport->m_process_response_callback(message, nullptr);
+
+                    if (!cts.get_token().is_canceled())
+                    {
+                        transport->receive_loop(cts);
```

Comment:
Could this stack dive if enough messages were buffered in memory that `websocket_client->receive` repeatedly called its callback inline?

Reply:
Definitely could. I believe we talked about this and decided to do it this way for now and re-evaluate later if we need to dispatch.

Reply:
If there's a risk, I figure it's easy enough to turn this into a trampoline to avoid the problem. I think it would look something like this.

```c++
auto calling_receive = true;
while (calling_receive) {
    websocket_client->receive([weak_transport, cts, logger, websocket_client, calling_receive](std::string message, std::exception_ptr exception)
        {
           // ...
           if (!cts.get_token().is_canceled())
           {
                if (calling_receive)
                {
                    calling_receive = false
                }
                else
                {
                    transport->receive_loop(cts);
                }
            }
        });

    calling_receive = !calling_receive;
}
```

Reply:
Something like that would work, but we still need to jump off the main thread for the start of the receive loop in order to avoid blocking start. Right now I'm relying on the transport dispatching, like we talked about in the meeting.

Reply:
What's the main thread? Whatever thread you call connection.start() on?

Do we jump off the main thread currently before calling websocket_client->receive()? Or does this design prevent us from jumping off the main thread in the future?

I would think the trampoline is strictly better than the recursive version because it avoids potentially stack diving.

Reply:
>What's the main thread? Whatever thread you call connection.start() on?

Correct.

>Do we jump off the main thread currently before calling websocket_client->receive()? Or does this design prevent us from jumping off the main thread in the future?

We don't do anything threading related yet, except in the http_client and websocket_client as those use pplx. We can add it whenever though.

>I would think the trampoline is strictly better than the recursive version because it avoids potentially stack diving.

I agree.